[ConstraintSystem] InferSendableFromCaptures: Mark unapplied operator references as `@Sendable`

xedin opened the pull request:
Operator references don't have a base type and could be found via
unqualified lookup, `adjustFunctionTypeForConcurrency` should handle
them specifically.

Resolves: rdar://131321053

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Fix `isPartialApplication` to handle unappled operator references

xedin committed:
[ConstraintSystem] InferSendableFromCaptures: Mark unapplied operator references as `@Sendable`

Both of these functions are single-apply instance members,
even though their names could be compound the application
is still a regular call with like `next(isolation: ...)`
vs. `next(isolation:)(...)` it would have to be for a "compound"
application.

Operator references don't have a base type and could be found via
unqualified lookup, `adjustFunctionTypeForConcurrency` should handle
them specifically.

Resolves: rdar://131321053

xedin commented:
Apparently expressions like `let _: @Sendable (Int, Int) -> Int = (+)` crash in verifier today because `ExprRewriter` doesn't inject `FunctionConversionExpr` for that. I'll fix that separately.

xedin commented:
@swift-ci please test

hamishknight commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -2644,6 +2647,11 @@ bool ConstraintSystem::isPartialApplication(ConstraintLocator *locator) {
       locator->findFirst<LocatorPathElt::ImplicitConversion>())
     return false;
 
+  if (locator->directlyAt<OverloadedDeclRefExpr>()) {
+    auto *ODRE = castToExpr<OverloadedDeclRefExpr>(locator->getAnchor());
+    return ODRE->getFunctionRefKind() == FunctionRefKind::Unapplied;
+  }
```
Should this also apply to regular `DeclRefExpr`s? Although `adjustFunctionTypeForConcurrency` already has `numApplies`, I wonder if `isPartialApplication` in this case should really be `numApplies < decl->getNumCurryLevels()` since `numApplies` should already account for currying? I should also note that it looks like the current logic here doesn't account for cases such as:

```swift
struct S {
  func foo() {}
}

func bar(_ x: @Sendable () -> Void) {}

let fn = S.foo(S())
bar(fn)
```

xedin replied:
Yeah, that sounds simpler, I'll give it a try!

xedin replied:
Done! We had function ref kind set incorrectly in a few places for synthesized code.

xedin committed:
[CSGen] Set correct ref kinds on synthesized `makeIterator` and `next` for for-in loops

xedin committed:
[ConstraintSystem] Use `getNumApplications` + `ValueDecl::getNumCurryLevels` to detect partial applications

We used to detect partial applications based on member locators
and parent expressions, but using function reference kind +
check to see if self is applied is such simpler and hits both
uses of `isPartialApplication`.

xedin committed:
[ConstraintSystem] NFC: Remove obsolete `isPartialApplication`

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

hamishknight reviewed (approved):
Looks great, thank you!

hamishknight commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -1774,9 +1774,12 @@ FunctionType *ConstraintSystem::adjustFunctionTypeForConcurrency(
+      } else if (numApplies < decl->getNumCurryLevels()) {
+        // Operators on protocols could be found via unqualified lookup and
+        // won't have a base type.
+        if (decl->isOperator() ||
+            (baseType &&
+             (baseType->is<AnyMetatypeType>() || baseType->isSendableType()))) {
```
Actually wait I think we need to check more than `baseType` for the curried instance method with applied self case, i.e for `X.foo(X())`, the base being a metatype is always true, but what matters is the Sendability of the instance type

xedin replied:
You are right! I'm working on a fix!

xedin replied:
✅

xedin committed:
[ConstraintSystem] InferSenableFromCaptures: Fix treatment of partially applied instance methods on non-Sendable base

In situations like:

```
class A {
  func test() {}
}
```

Fully uncurried function type should be @Sendable but inner method
cannot be because it captures a non-Sendable type `A` when applied.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility
